dipy/segment/fss.py

```diff
@@ -0,0 +1,291 @@
+
+import warnings
+import numpy as np
+from scipy.sparse import coo_matrix
+from scipy.spatial import cKDTree
+
+from dipy.tracking.streamline import set_number_of_points
+from dipy.segment.metric import mean_l2_func
+
+
+class FastStreamlineSearch:
+    def __init__(self, ref_streamlines, max_radius, nb_mpts=4, bin_size=20.0,
+                 resampling=24, bidirectional=True):
+        """ Fast Streamline Search (FFS)
+
+        Generate the Binned K-D Tree structure with reference streamlines,
+        using streamlines barycenter and mean-points.
+        See [StOnge2021]_ for further details.
+
+        Parameters
+        ----------
+        ref_streamlines : Streamlines
+            Streamlines (ref) to generate the tree structure.
+        max_radius : float
+            The maximum radius (distance) for subsequent streamline search.
+            Used to compute the overlap in-between bins.
+        nb_mpts : int, optional
+            Number of means points to improve computation speed.
+            (this only changes computation time)
+        bin_size : float, optional
+            The bin size to separate streamlines in groups.
+            (this only changes computation time)
+        resampling : int, optional
+            Number of points used to reshape each streamline.
+        bidirectional : bool, optional
+            Compute the smallest distance with and without flip.
+
+        Notes
+        -----
+        Make sure that streamlines are aligned in the same space.
+        Preferably in millimeter space (voxmm or rasmm).
+
+        References
+        ----------
+        .. [StOnge2021] St-Onge E. et al. Fast Tractography Streamline Search,
+                        International Workshop on Computational Diffusion MRI,
+                        pp. 82-95. Springer, Cham, 2021.
+        """
+        if max_radius <= 0.0:
+            raise ValueError("max_radius needs to be a positive value")
+
+        if resampling < 20:
+            warnings.warn("For accurate results, resampling should be"
+                          " at least >= 10 and preferably >= 20")
+
+        if resampling % nb_mpts != 0:
+            raise ValueError("nb_mpts needs to be a factor of resampling")
```

Comment:
It's not really a speedup factor, since it could increase the computation time.
However, it could be a "simplification_factor", but refered as mean-points in the paper.

It also needs to be an exact divisor ex:  resampling = 3 can only have 1 or 3 mean-points
so it wont avoid the check by using a int division "//"